src/fft.cpp

```diff
@@ -7,106 +7,120 @@ using namespace std;
       auto sub1   = ducc0::subarray(data, {{}, {}, {}, {0, z_lo}});
       auto sub2   = ducc0::subarray(data, {{}, {}, {}, {z_hi, ducc0::MAXIDX}});
       auto sub3   = ducc0::subarray(sub1, {{}, {}, {0, y_lo}, {}});
       auto sub4   = ducc0::subarray(sub1, {{}, {}, {y_hi, ducc0::MAXIDX}, {}});
       auto sub5   = ducc0::subarray(sub2, {{}, {}, {0, y_lo}, {}});
       auto sub6   = ducc0::subarray(sub2, {{}, {}, {y_hi, ducc0::MAXIDX}, {}});
-      if (p->type == 1) { // spreading, not all parts of the output array are needed
+      if (spreading) { // spreading, not all parts of the output array are needed
         // do axis 3 in full
-        ducc0::c2c(data, data, {3}, p->fftSign < 0, TF(1), nthreads);
+        ducc0::c2c(data, data, {3}, forward, TF(1), nthreads);
         // do only parts of axis 2
-        ducc0::c2c(sub1, sub1, {2}, p->fftSign < 0, TF(1), nthreads);
-        ducc0::c2c(sub2, sub2, {2}, p->fftSign < 0, TF(1), nthreads);
+        ducc0::c2c(sub1, sub1, {2}, forward, TF(1), nthreads);
+        ducc0::c2c(sub2, sub2, {2}, forward, TF(1), nthreads);
       }
       // do even smaller parts of axis 1
-      ducc0::c2c(sub3, sub3, {1}, p->fftSign < 0, TF(1), nthreads);
-      ducc0::c2c(sub4, sub4, {1}, p->fftSign < 0, TF(1), nthreads);
-      ducc0::c2c(sub5, sub5, {1}, p->fftSign < 0, TF(1), nthreads);
-      ducc0::c2c(sub6, sub6, {1}, p->fftSign < 0, TF(1), nthreads);
-      if (p->type == 2) { // interpolation, parts of the input array are zero
+      ducc0::c2c(sub3, sub3, {1}, forward, TF(1), nthreads);
+      ducc0::c2c(sub4, sub4, {1}, forward, TF(1), nthreads);
+      ducc0::c2c(sub5, sub5, {1}, forward, TF(1), nthreads);
+      ducc0::c2c(sub6, sub6, {1}, forward, TF(1), nthreads);
+      if (!spreading) { // interpolation, parts of the input array are zero
         // do only parts of axis 2
-        ducc0::c2c(sub1, sub1, {2}, p->fftSign < 0, TF(1), nthreads);
-        ducc0::c2c(sub2, sub2, {2}, p->fftSign < 0, TF(1), nthreads);
+        ducc0::c2c(sub1, sub1, {2}, forward, TF(1), nthreads);
+        ducc0::c2c(sub2, sub2, {2}, forward, TF(1), nthreads);
         // do axis 3 in full
-        ducc0::c2c(data, data, {3}, p->fftSign < 0, TF(1), nthreads);
+        ducc0::c2c(data, data, {3}, forward, TF(1), nthreads);
       }
     }
   }
 #endif
 #else
-  p->fftPlan->execute(); // if thisBatchSize<batchSize it wastes some flops
+  // FIXME: the "adjoint" emulation is a crude band-aid
+  if (adjoint)
+    for (BIGINT i = 0; i < ntrans_actual * p.nf(); ++i) fwBatch[i] = conj(fwBatch[i]);
+  p.fftPlan->execute(fwBatch); // if thisBatchSize<batchSize it wastes some flops
+  if (adjoint)
+    for (BIGINT i = 0; i < ntrans_actual * p.nf(); ++i) fwBatch[i] = conj(fwBatch[i]);
 #endif
 }
-template void do_fft<float>(FINUFFT_PLAN_T<float> *p);
-template void do_fft<double>(FINUFFT_PLAN_T<double> *p);
+template void do_fft<float>(const FINUFFT_PLAN_T<float> &p, std::complex<float> *fwBatch,
```

Comment:
Is there an FFTW version of passing the data ptr like this?

Reply:
https://fftw.org/fftw3_doc/New_002darray-Execute-Functions.html#New_002darray-Execute-Functions

Reply:
The new adjoint FFTW is excellent. THanks!